### DISCUSSION TITLE
Milvus FAISS Parallel Queries on top of GPU

### DISCUSSION THREAD

**Discussion by astar10239** [milvus-io/milvus#5532] [Q&A and General discussion]
Hi,

I am looking for a implementation of MILVUS/FAISS Over GPU

I have a very large index of 100 million embeddings and want real-time search on api

Can anyone help here,

Does Milvus/Faiss Support Parallel Queries over GPU or at-least Batch queries over GPU (given each query is different)

**Comment by yhmo** [comment]
GPU is faster than cpu for intensive computing. The limitation is gpu memory. Currently, for most of gpu devices, gpu memory is 4GB ~ 8GB. For 100M embeddings, if the dimension is 256, total data size is 100GB, even with some index(SQ,PQ) that compress embedding, the index size still more than 20GB. That means most of GPU device cannot hold large dataset in gpu memory completely.

Milvus split large dataset into pieces, we called "segment", you can refer this page: https://milvus.io/docs/v1.1.0/storage_concept.md
Typically, each segment size is 1GB ~ 4GB, according to users requirement. When mivus get a search request, it search for each segment one by one, each segment output a topk result, then milvus reduce multiple results into a final topk result. 
By this machinery, milvus can use different devices to help searching. For example, there are 4 GPU devices on the machine, milvus can send segment_1 to gpu_1, segment_2 to gpu_2, ... These devices can compute at the same time.

Because gpu computing need to copy data from cpu memory to gpu memory, there is time cost in copying data, although gpu parallel compute ability is stronger than cpu, the performance may not be good for small batch query. Large batch query is suitable for GPU.

Milvus configuration support specify multiple GPU devices for search, in the server_config.yaml:
```
gpu:
  enable: true
  cache_size: 2GB
  gpu_search_threshold: 1000
  search_devices:
    - gpu1,gpu2,gpu3
  build_index_devices:
    - gpu0
```

**Reply by astar10239** [reply]
Hi,

Thank you for the reply.

Understood.

I have a 32GB GPU V100

What i am observing is milvus is loading it into GPU and removing the Index after the query, cant it be kept on memory of the GPU so that latency can be saved ?

What is I make the segment size more then 8GB or something so that in one segment maximum search can be done.

Also does milvus support parallel queries.

Actually when i benchmarked i found out that milvus was supporting parallel but i tried only 10 million embeddings but it was taking 7-9 sec.

Observation was it was trying load-deload the index on GPU with every search request

**Reply by yhmo** [reply]
Yes, in v1.1.0 and previous versions, milvus always reload data from cpu memory into gpu memory, because we think most users have multiple segments in one segment, and most of gpu memory only can cache one segment for searching, search engine need to load another segment after the previous finish searching. 

Some issues require cache data in gpu memory: #1954, #5142
In next minor version, we decide to support cache segment data in gpu memory, this has been implemented by the pr #5161.

For "segment size more then 8GB", currently there is a limitation "4GB", to break the limitation, some user change the source code and compile milvus by themself. Personally I think "4GB" need to be enlarge.

**Reply by astar10239** [reply]
Ok,

I see that in 1.1 its implemented

gpu:
  cache.enable: false

Do you have the docker image for this version ?

Is this the one ?

https://hub.docker.com/layers/milvusdb/milvus/1.1.0-gpu-d050721-5e559c/images/sha256-4bf7ef0bed7cae4be8374308c980d15abd263a0ddbde10871662206570cb1e26?context=explore

If not then may i know when is next release with the support.

Also when you cache the segments in GPU

Is it like let say a GPU has 32GB memory, so you will cache max 32/4 number of indexes, or only 1 index at a time of max 4gb ?

**Reply by yhmo** [reply]
The link you give is v1.1.0 which was published on May 07,2021. 
The next minor version is v1.1.1, we are testing it, I think one week is required.

The next version will set cache size according to the configuration:
Assume your gpu has 32GB memory, you set cache size =20GB in the config file, each index is 4GB, then it can hold 5 index in gpu memory.
```
gpu:
  enable: true
  cache_size: 20GB
  gpu_search_threshold: 1000
  search_devices:
    - gpu1,gpu2,gpu3
  build_index_devices:
    - gpu0
```
And there is a checker during milvus startup, the gpu_cache_size must less than your_gpu_memory * 0.9.
Personally I think the "0.9" is too risk, because the search engine may require extra gpu memory during searching, maybe "0.6" is safe.

**Reply by astar10239** [reply]
Great, thank you waiting for you next version v1.1.1

So v1.1.1 will have support for gpu index cache and cache size changes right
